js/modules/k6/experimental/fs/module.go

```diff
@@ -165,3 +170,102 @@ func (f *File) Stat() *goja.Promise {
 
 	return promise
 }
+
+// Read the file's content, and writes it into the provided Uint8Array.
+//
+// Resolves to either the number of bytes read during the operation
+// or EOF (null) if there was nothing more to read.
+//
+// It is possible for a read to successfully return with 0 bytes.
+// This does not indicate EOF.
+func (f *File) Read(into goja.Value) *goja.Promise {
+	promise, resolve, reject := promises.New(f.vu)
+
+	if common.IsNullish(into) {
+		reject(newFsError(TypeError, "read() failed; reason: into cannot be null or undefined"))
+		return promise
+	}
+
+	// We expect the into argument to be a `Uint8Array` instance
+	intoObj := into.ToObject(f.vu.Runtime())
+	uint8ArrayConstructor := f.vu.Runtime().Get("Uint8Array")
+	if isUint8Array := intoObj.Get("constructor").SameAs(uint8ArrayConstructor); !isUint8Array {
+		reject(newFsError(TypeError, "read() failed; reason: into argument must be a Uint8Array"))
+		return promise
+	}
+
+	// Obtain the underlying ArrayBuffer from the Uint8Array
+	ab, ok := intoObj.Get("buffer").Export().(goja.ArrayBuffer)
+	if !ok {
+		reject(newFsError(TypeError, "read() failed; reason: into argument cannot be interpreted as ArrayBuffer"))
+		return promise
+	}
+
+	// Obtain the underlying byte slice from the ArrayBuffer.
+	// Note that this is not a copy, and will be modified by the Read operation
+	// in place.
+	buffer := ab.Bytes()
+
+	go func() {
+		n, err := f.file.Read(buffer)
+		if err != nil {
```

Comment:
nit: I think this is relevant here as this `Read` is not the same as [os.File#Read](https://pkg.go.dev/os#File.Read) but it can both have read stuff *and*  returned data. 

[`io.Reader`](https://pkg.go.dev/io#Reader) does define this as a behaviour that can be expected.

I am not particularly certain how to handle this in js as  if we did read some amount and then got EOF and `Read` returns both we will hit the bleow and resolve with `nil`. Which likely isn't great.

At this point this is less of a concern as the below implementation is specifically implemented by us. But in a refactor this might become a problem. 

A possible solution could be to retrurn this buffer and remember that there was an error and return it the next time :shrug:

Reply:
Hey, I'm not sure I understand what you mean here. Especially when it comes to "it can both have read stuff and returned data.".

The reason I don't understand is likely that Read modifies the buffer in place (`buffer` is a pointer and not a copy, as I learned the hard way in WebCrypto), and returns either a number or `null` to indicate EOF.

Could you elaborate, or even better provide a JS example to illustrate what you meant? That would be super helpful 🙇🏻

Reply:
> returns either a number or null to indicate EOF

This is true, but `io.Reader` and `os.File#Read` *can* return both `5` and `EOF` at the same time - which means that you read 5 bytes in the array and got `EOF`.

The current code will though resolve with `null` which means that a user of the js code will have no idea taht actually there were 5 bytes read. Yes they are in the ArrayBuffer - true, but you don't know that there are 5 new bytes, just that you got EOF. 

Even if for EOF you return `5`, this can still happen with some other error. In which case it won't work so well.

> or even better provide a JS example to illustrate what you meant? 

The problem here is that JS does not have multiple returns as go does. 

Again as I mention currently the underlying implementation would not present this problem as .... we can't have anyu errors really apart from EOF and it happens only when it returns 0 bytes . Which is why this is a  `nit`.

But in the future if this changes it might become a problem.

Hope this explained it more?

Reply:
> Hope this explained it more?

Absolutely, thank you very much 🙇🏻 

Makes sense, I had completely overlooked that case, indeed! I am going to look into how to cover that scenario indeed, I find it worth it 👀

Reply:
I aligned the behavior of `file.Read` with `io.Reader.Read`'s, adjusted the module code, and tested accordingly in [`3000972` (#3309)](https://github.com/grafana/k6/pull/3309/commits/3000972bdc4f085567e023128c894ce56a00ff1e). Let me know if you think that addresses. fully your point 🙇🏻